src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            var newName = newDocumentInfo.Name;
+            var shouldRenameType = !string.Equals(document.Name, newName, StringComparison.OrdinalIgnoreCase);
+            var shouldMoveNamespace = !document.Folders.SequenceEqual(newDocumentInfo.Folders);
+
+            var solution = document.Project.Solution
+                .WithDocumentName(document.Id, newDocumentInfo.Name)
+                .WithDocumentFilePath(document.Id, newDocumentInfo.FilePath)
+                .WithDocumentFolders(document.Id, newDocumentInfo.Folders);
+
+            if (!(shouldRenameType || shouldMoveNamespace))
+            {
+                return new RenameDocumentResult(solution, renameSymbolSucceeded: false);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (shouldRenameType)
+            {
+                solution = await RenameTypeToMatchDocumentNameAsync(document.Id, solution, newName, document.Name, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (shouldMoveNamespace)
+            {
+                solution = await SyncNamespaceToDocumentFoldersAsync(document.Id, solution, cancellationToken).ConfigureAwait(false);
+            }
+
+            return new RenameDocumentResult(solution, renameSymbolSucceeded: true);
+        }
+
+        private static async Task<Solution> RenameTypeToMatchDocumentNameAsync(DocumentId documentId, Solution solution, string newName, string originalName, CancellationToken cancellationToken)
+        {
+            var newNameWithoutExtension = Path.GetFileNameWithoutExtension(newName);
+            var oldNameWithoutExtension = Path.GetFileNameWithoutExtension(originalName);
+
+            var document = solution.GetDocument(documentId);
+
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
+            var displayNames = typeDeclarations.Select(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None));
+
+            var matchingTypeDeclaration = typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(oldNameWithoutExtension, StringComparison.OrdinalIgnoreCase));
+            if (matchingTypeDeclaration is object)
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                solution = await RenameSymbolAsync(solution, symbol, newNameWithoutExtension, solution.Options, cancellationToken).ConfigureAwait(false);
+            }
+
+            return solution;
```

Comment:
this is weird.  so if we don't find a matchin type, we just return the solution unchanged.  but the caller says "rename succeeded".   feels like we should let the caller know "we didn't actually rename anything because we couldn't find a type that matched"

Reply:
I think this should go into the discussion about providing more information than a single success/failure bool. That would help with this scenario as well.